src/Week3/Fun.hs

```diff
@@ -0,0 +1,78 @@
+module Fun where
+
+data Expr = Lit    Bool
+          | Var    String
+          | Let    String Expr Expr
+          | If     Expr Expr Expr
+          | Lambda String Expr
+          | App    Expr Expr
+            deriving (Eq)
+
+
+instance Show Expr where
```

Comment:
This Show instance is fine for playing around, but be aware that it breaks Show contract.

The general requirements for Show instances are described [here](https://hackage.haskell.org/package/base-4.15.0.0/docs/Prelude.html#t:Show). One of the requirements is that Show instance (which converts you type to String) should be compatible with Read (which converts a String back to a value of your type). So calling `read (show someValue)` should return `someValue`.

So what you want to do is usually to use `deriving (Show)` and to get nice textual representation of your values for printing you'd usually use some pretty printer library (like [prettyprinter](https://hackage.haskell.org/package/prettyprinter)). But that is beyond the scope of these exercises. Mentioning it just so you're aware.